archive/brotli.cpp

```diff
@@ -0,0 +1,101 @@
+// SPDX-FileCopyrightText: 2024 David Zero <zero-one@zer0-one.net>
+//
+// SPDX-License-Identifier: BSD-2-Clause
+
+#include "archive/brotli.h"
+
+#include "brotli/decode.h"
+#include <tl/expected.hpp>
+
+#include <cstddef>
+#include <cstdint>
+#include <memory>
+#include <span>
+#include <string_view>
+#include <vector>
+
+namespace archive {
+
+std::string_view to_string(BrotliError err) {
+    switch (err) {
+        case BrotliError::DecoderState:
+            return "Failed to create brotli decoder state";
+        case BrotliError::InputCorrupt:
+            return "Input is corrupt or truncated";
+        case BrotliError::InputEmpty:
+            return "Input is empty";
+        case BrotliError::MaximumOutputLengthExceeded:
+            return "Output buffer exceeded maximum allowed length";
+        case BrotliError::BrotliInternalError:
+            return "Decode failure";
+    }
+
+    return "Unknown error";
+}
+
+tl::expected<std::vector<std::byte>, BrotliError> brotli_decode(std::span<std::byte const> const input) {
+    if (input.empty()) {
+        return tl::unexpected{BrotliError::InputEmpty};
+    }
+
+    std::unique_ptr<BrotliDecoderState, decltype(&BrotliDecoderDestroyInstance)> br_state(
+            BrotliDecoderCreateInstance(nullptr, nullptr, nullptr), BrotliDecoderDestroyInstance);
+
+    if (br_state == nullptr) {
+        return tl::unexpected{BrotliError::DecoderState};
+    }
+
+    // Cap output buffer at 1GB. If we hit this, something fishy is probably
+    // going on, and we should bail before we OOM.
+    std::size_t constexpr kMaxOutSize = 1000000000;
+    std::size_t constexpr kChunkSize = 131072; // Matches the zstd chunk size
+
+    std::vector<std::byte> out;
+
+    std::size_t avail_in = input.size();
+    auto const *next_in = reinterpret_cast<std::uint8_t const *>(input.data());
+    std::size_t total_out = 0;
+
+    BrotliDecoderResult res = BROTLI_DECODER_RESULT_ERROR;
+
+    while (res != BROTLI_DECODER_RESULT_SUCCESS) {
+        std::vector<std::byte> intermediate_buf(kChunkSize);
+
+        std::size_t avail_out = kChunkSize;
+        auto *next_out = reinterpret_cast<std::uint8_t *>(intermediate_buf.data());
+
+        if (out.size() >= kMaxOutSize) {
+            return tl::unexpected{BrotliError::MaximumOutputLengthExceeded};
+        }
+
+        res = BrotliDecoderDecompressStream(br_state.get(), &avail_in, &next_in, &avail_out, &next_out, &total_out);
+
+        // Because we provide the whole input up-front, there's no reason we
+        // would ever block on needing more input, except for corrupt data
+        if (res == BROTLI_DECODER_RESULT_NEEDS_MORE_INPUT) {
+            return tl::unexpected{BrotliError::InputCorrupt};
```

Comment:
InputIncomplete is the correct error here

Reply:
Incomplete input is corrupt input.

Reply:
They're unique states. In one, more data would fix the problem, and allow it to continue, in the other more data won't fix the problem.